src/hx/libs/regexp/RegExp.cpp

```diff
@@ -34,76 +43,92 @@ struct pcredata : public hx::Object
       HX_OBJ_WB_GET(this, expr.raw_ref());
       flags = inFlags;
 
-
-      nmatchs = 0;
-      pcre_fullinfo(rUtf8,NULL,PCRE_INFO_CAPTURECOUNT,&nmatchs);
-      nmatchs++;
-      matchs = (int*)malloc(sizeof(int) * 3 * nmatchs);
+      n_groups = 0;
+      pcre2_pattern_info_8(rUtf8,PCRE2_INFO_CAPTURECOUNT,&n_groups);
+      n_groups++;
+      match_data8 = pcre2_match_data_create_from_pattern_8(rUtf8, NULL);
+      #ifdef HX_SMART_STRINGS
+      match_data16 = 0;
+      #endif
 
       _hx_set_finalizer(this, finalize);
    }
 
    #ifdef HX_SMART_STRINGS
-   void create16(pcre16 *inR, String inExpr, int inFlags)
+   void create16(pcre2_code_16 *inR, String inExpr, int inFlags)
    {
       rUtf8 = 0;
       rUtf16 = inR;
       expr = inExpr;
       HX_OBJ_WB_GET(this, expr.raw_ref());
       flags = inFlags;
 
-      nmatchs = 0;
-      pcre16_fullinfo(rUtf16,NULL,PCRE_INFO_CAPTURECOUNT,&nmatchs);
-      nmatchs++;
-      matchs = (int*)malloc(sizeof(int) * 3 * nmatchs);
+      n_groups = 0;
+      pcre2_pattern_info_16(rUtf16,PCRE2_INFO_CAPTURECOUNT,&n_groups);
+      n_groups++;
+      match_data8 = 0;
+      match_data16 = pcre2_match_data_create_from_pattern_16(rUtf16, NULL);
 
       _hx_set_finalizer(this, finalize);
    }
    #endif
 
-
    bool run(String string,int pos,int len)
    {
       #ifdef HX_SMART_STRINGS
       if (string.isUTF16Encoded())
       {
          if (!rUtf16)
          {
-            const char *error = 0;
-            int err_offset = 0;
+            int error_code;
+            size_t error_offset;
             hx::strbuf buf;
-            rUtf16 = pcre16_compile((PCRE_SPTR16)expr.wc_str(&buf),flags|PCRE_UTF16,&error,&err_offset,NULL);
-            if (!rUtf16)
-            {
-               return false;
+            rUtf16 = pcre2_compile_16((PCRE2_SPTR16)expr.wc_str(&buf),expr.length,flags,&error_code,&error_offset,NULL);

```

Comment:
Is `expr.length` in bytes or code units? The byte length is not the same thing as code unit length in both UTF-8 and UTF-16. UTF-8 is a multi-byte encoding and UTF-16 (with surrogates) is a multi-word16 encoding.

Reply:
Ah, yes actually we need to get the new length here as well after converting to utf-16

Reply:
Hm, when hxcpp_smart_strings is enabled, utf-8 is only ever used for non-unicode strings, and hxcpp seems pretty good at making sure that utf-16 is used as soon as any unicode characters are introduced. So it looks like it might be impossible to have a utf-8 conversion here where it is actually using any multi byte characters, and so the length will always be the same... We can "fix" this by adding a byteLength input parameter to `wc_str()`, but there isn't an easy way to create a test for this.

Reply:
I cannot say I know much about hxcpp or its smart string feature but I am confused by your statement "utf-8 is only ever used for non-unicode strings" since UTF-8 is a Unicode encoding.

Also "hxcpp seems pretty good at making sure that utf-16 is used as soon as any unicode characters are introduced" is hard to understand too. Which characters are not Unicode?

Reply:
With hxcpp_smart_strings, if a string only contains characters that can be represented by ASCII, it will use 8 bit encoding to save space. If it uses any non-ASCII characters then it uses utf-16. So in practice, we are not really converting utf-8 to utf-16, but ASCII to utf-16, so the length is going to remain consistent. 
Anyway, in the latest commits I fixed it so it will use the updated utf-16 length just in case, but it won't make a difference really.